checker.cpp

```diff
@@ -1,22 +1,25 @@
 #include <assert.h>
 #include <iostream>
+#include <memory>
+#include "checker.hpp"
+#include "BMSParam.hpp"
+#include "MessageProvider.hpp"
 using namespace std;
 
-bool batteryIsOk(float temperature, float soc, float chargeRate) {
-  if(temperature < 0 || temperature > 45) {
-    cout << "Temperature out of range!\n";
-    return false;
-  } else if(soc < 20 || soc > 80) {
-    cout << "State of Charge out of range!\n";
-    return false;
-  } else if(chargeRate > 0.8) {
-    cout << "Charge Rate out of range!\n";
-    return false;
-  }
-  return true;
+bool batteryIsApproachingThreshold(float temperature, float soc, float chargeRate)
+{
+  std::shared_ptr<MessageProvider> engMsgProvider(new EnglishMessageProvider()); 
+  std::shared_ptr<BMSParam> temp(new Temperature(temperature, engMsgProvider.get())); 
+  std::shared_ptr<BMSParam> stateOfCharge(new SOC(soc, engMsgProvider.get()));
+  std::shared_ptr<BMSParam> rateOfCharge(new ChargeRate(chargeRate, engMsgProvider.get()));
+  return (temp->checkIfToleranceApproaching() || stateOfCharge->checkIfToleranceApproaching() || rateOfCharge->checkIfToleranceApproaching());
 }
 
-int main() {
-  assert(batteryIsOk(25, 70, 0.7) == true);
-  assert(batteryIsOk(50, 85, 0) == false);
+bool batteryIsOk(float temperature, float soc, float chargeRate)
+{
+  std::shared_ptr<MessageProvider> grMsgProvider(new GermanMessageProvider()); 
```

Comment:
The instantiation of the message provider should not be done by the consumer. It should be created outside and injected to the client. 
In this case if non German message provider needs to be used, then you have to modify code, compile and deploy.
You already have all the necessary code to dynamically create the required message provider.

In other words, the message provide should have been created in main().